File: volume_calculator.go
// volume_calculator.go - Калькулятор объёма тел на Go (CLI)
package main

import (
	"bufio"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
	"time"
)

const PI = math.Pi
const HISTORY_FILE = "volume_history.json"

type Shape struct {
	Name   string
	Func   func([]float64) float64
	Params []string
}

type HistoryEntry struct {
	Date   string `json:"date"`
	Shape  string `json:"shape"`
	Params string `json:"params"`
	Result string `json:"result"`
	Units  string `json:"units"`
}

var shapes = map[string]Shape{
	"1":  {"Куб", func(p []float64) float64 { return math.Pow(p[0], 3) }, []string{"сторону"}},
	"2":  {"Шар", func(p []float64) float64 { return 4.0 / 3.0 * PI * math.Pow(p[0], 3) }, []string{"радиус"}},
	"3":  {"Цилиндр", func(p []float64) float64 { return PI * math.Pow(p[0], 2) * p[1] }, []string{"радиус", "высоту"}},
	"4":  {"Конус", func(p []float64) float64 { return 1.0 / 3.0 * PI * math.Pow(p[0], 2) * p[1] }, []string{"радиус", "высоту"}},
	"5":  {"Параллелепипед", func(p []float64) float64 { return p[0] * p[1] * p[2] }, []string{"длину", "ширину", "высоту"}},
	"6":  {"Пирамида", func(p []float64) float64 { return 1.0 / 3.0 * p[0] * p[1] }, []string{"площадь основания", "высоту"}},
	"7":  {"Призма", func(p []float64) float64 { return p[0] * p[1] }, []string{"площадь основания", "высоту"}},
	"8":  {"Тор", func(p []float64) float64 { return 2.0 * math.Pow(PI, 2) * p[0] * math.Pow(p[1], 2) }, []string{"большой радиус", "малый радиус"}},
	"9":  {"Эллипсоид", func(p []float64) float64 { return 4.0 / 3.0 * PI * p[0] * p[1] * p[2] }, []string{"полуось a", "полуось b", "полуось c"}},
	"10": {"Усечённый конус", func(p []float64) float64 { return 1.0 / 3.0 * PI * p[2] * (math.Pow(p[0], 2) + p[0]*p[1] + math.Pow(p[1], 2)) }, []string{"радиус 1", "радиус 2", "высоту"}},
}

func saveHistory(entry HistoryEntry) {
	var history []HistoryEntry
	file, err := os.ReadFile(HISTORY_FILE)
	if err == nil {
		json.Unmarshal(file, &history)
	}
	history = append(history, entry)
	data, _ := json.MarshalIndent(history, "", "  ")
	os.WriteFile(HISTORY_FILE, data, 0644)
}

func loadHistory() []HistoryEntry {
	var history []HistoryEntry
	file, err := os.ReadFile(HISTORY_FILE)
	if err != nil {
		return history
	}
	json.Unmarshal(file, &history)
	return history
}

func exportCSV(filename string) {
	history := loadHistory()
	if len(history) == 0 {
		fmt.Println("История пуста.")
		return
	}
	f, _ := os.Create(filename)
	defer f.Close()
	writer := csv.NewWriter(f)
	defer writer.Flush()
	writer.Write([]string{"Дата", "Фигура", "Параметры", "Объём", "Единицы"})
	for _, e := range history {
		units := e.Units
		if units == "" {
			units = "куб. ед."
		}
		writer.Write([]string{e.Date[:19], e.Shape, e.Params, e.Result, units})
	}
	fmt.Printf("Экспортировано в %s\n", filename)
}

func getFloat(prompt string, reader *bufio.Reader) float64 {
	for {
		fmt.Print(prompt)
		input, _ := reader.ReadString('\n')
		input = strings.TrimSpace(input)
		val, err := strconv.ParseFloat(input, 64)
		if err == nil && val > 0 {
			return val
		}
		fmt.Println("Введите положительное число.")
	}
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	fmt.Println("📦 КАЛЬКУЛЯТОР ОБЪЁМА ТЕЛ")
	history := loadHistory()
	units := "куб. ед."

	for {
		fmt.Println("\nВыберите фигуру:")
		for key, shape := range shapes {
			fmt.Printf("%s. %s\n", key, shape.Name)
		}
		fmt.Println("h. Показать историю")
		fmt.Println("e. Экспорт CSV")
		fmt.Println("u. Сменить единицы")
		fmt.Println("0. Выход")

		fmt.Print("Ваш выбор: ")
		choice, _ := reader.ReadString('\n')
		choice = strings.TrimSpace(choice)

		if choice == "0" {
			break
		} else if choice == "h" || choice == "H" {
			if len(history) == 0 {
				fmt.Println("История пуста.")
			} else {
				fmt.Println("\n=== ИСТОРИЯ ===")
				start := len(history) - 10
				if start < 0 {
					start = 0
				}
				for i := start; i < len(history); i++ {
					e := history[i]
					fmt.Printf("%s | %s | %s\n", e.Date[:19], e.Shape, e.Result)
				}
			}
			continue
		} else if choice == "e" || choice == "E" {
			fmt.Print("Имя CSV файла (по умолчанию volume_history.csv): ")
			filename, _ := reader.ReadString('\n')
			filename = strings.TrimSpace(filename)
			if filename == "" {
				filename = "volume_history.csv"
			}
			exportCSV(filename)
			continue
		} else if choice == "u" || choice == "U" {
			fmt.Println("Доступные единицы: куб. ед., см³, м³, дм³, л, in³")
			fmt.Print("Выберите единицы: ")
			input, _ := reader.ReadString('\n')
			input = strings.TrimSpace(input)
			if input != "" {
				units = input
			}
			continue
		} else if shape, ok := shapes[choice]; ok {
			fmt.Printf("\nФигура: %s\n", shape.Name)
			params := make([]float64, len(shape.Params))
			for i, pname := range shape.Params {
				params[i] = getFloat("Введите "+pname+": ", reader)
			}
			result := shape.Func(params)
			resultStr := fmt.Sprintf("%.4f %s", result, units)
			fmt.Printf("\nОбъём %s: %s\n", strings.ToLower(shape.Name), resultStr)

			fmt.Print("Сохранить результат? (y/n): ")
			saveChoice, _ := reader.ReadString('\n')
			saveChoice = strings.TrimSpace(strings.ToLower(saveChoice))
			if saveChoice == "y" {
				entry := HistoryEntry{
					Date:   time.Now().Format(time.RFC3339),
					Shape:  shape.Name,
					Params: strings.Join(strings.Fields(fmt.Sprint(params)), ", "),
					Result: resultStr,
					Units:  units,
				}
				saveHistory(entry)
				history = append(history, entry)
				fmt.Println("✅ Сохранено!")
			}
		} else {
			fmt.Println("Неверный выбор.")
		}
	}
}

File: volume_calculator.rs
// volume_calculator.rs - Калькулятор объёма тел на Rust (CLI)
use std::collections::HashMap;
use std::fs;
use std::io::{self, Write};
use std::str::FromStr;

const PI: f64 = std::f64::consts::PI;
const HISTORY_FILE: &str = "volume_history.json";

#[derive(serde::Serialize, serde::Deserialize, Clone)]
struct HistoryEntry {
    date: String,
    shape: String,
    params: String,
    result: String,
    units: String,
}

type ShapeFunc = fn(Vec<f64>) -> f64;

fn get_shapes() -> HashMap<String, (String, ShapeFunc, Vec<String>)> {
    let mut m = HashMap::new();
    m.insert("1".to_string(), ("Куб".to_string(), |p| p[0].powf(3.0), vec!["сторону".to_string()]));
    m.insert("2".to_string(), ("Шар".to_string(), |p| 4.0/3.0 * PI * p[0].powf(3.0), vec!["радиус".to_string()]));
    m.insert("3".to_string(), ("Цилиндр".to_string(), |p| PI * p[0].powf(2.0) * p[1], vec!["радиус".to_string(), "высоту".to_string()]));
    m.insert("4".to_string(), ("Конус".to_string(), |p| 1.0/3.0 * PI * p[0].powf(2.0) * p[1], vec!["радиус".to_string(), "высоту".to_string()]));
    m.insert("5".to_string(), ("Параллелепипед".to_string(), |p| p[0] * p[1] * p[2], vec!["длину".to_string(), "ширину".to_string(), "высоту".to_string()]));
    m.insert("6".to_string(), ("Пирамида".to_string(), |p| 1.0/3.0 * p[0] * p[1], vec!["площадь основания".to_string(), "высоту".to_string()]));
    m.insert("7".to_string(), ("Призма".to_string(), |p| p[0] * p[1], vec!["площадь основания".to_string(), "высоту".to_string()]));
    m.insert("8".to_string(), ("Тор".to_string(), |p| 2.0 * PI.powf(2.0) * p[0] * p[1].powf(2.0), vec!["большой радиус".to_string(), "малый радиус".to_string()]));
    m.insert("9".to_string(), ("Эллипсоид".to_string(), |p| 4.0/3.0 * PI * p[0] * p[1] * p[2], vec!["полуось a".to_string(), "полуось b".to_string(), "полуось c".to_string()]));
    m.insert("10".to_string(), ("Усечённый конус".to_string(), |p| 1.0/3.0 * PI * p[2] * (p[0].powf(2.0) + p[0]*p[1] + p[1].powf(2.0)), vec!["радиус 1".to_string(), "радиус 2".to_string(), "высоту".to_string()]));
    m
}

fn save_history(entry: &HistoryEntry) {
    let mut history: Vec<HistoryEntry> = Vec::new();
    if let Ok(data) = fs::read_to_string(HISTORY_FILE) {
        if let Ok(parsed) = serde_json::from_str(&data) {
            history = parsed;
        }
    }
    history.push(entry.clone());
    let json = serde_json::to_string_pretty(&history).unwrap();
    fs::write(HISTORY_FILE, json).unwrap();
}

fn load_history() -> Vec<HistoryEntry> {
    if let Ok(data) = fs::read_to_string(HISTORY_FILE) {
        if let Ok(parsed) = serde_json::from_str(&data) {
            return parsed;
        }
    }
    Vec::new()
}

fn export_csv(filename: &str) {
    let history = load_history();
    if history.is_empty() {
        println!("История пуста.");
        return;
    }
    let mut content = String::from("Дата,Фигура,Параметры,Объём,Единицы\n");
    for e in &history {
        let units = if e.units.is_empty() { "куб. ед." } else { &e.units };
        content.push_str(&format!("{},{},{},{},{}\n", &e.date[..19], e.shape, e.params, e.result, units));
    }
    fs::write(filename, content).unwrap();
    println!("Экспортировано в {}", filename);
}

fn read_line(prompt: &str) -> String {
    print!("{}", prompt);
    io::stdout().flush().unwrap();
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    input.trim().to_string()
}

fn get_float(prompt: &str) -> f64 {
    loop {
        let input = read_line(prompt);
        if let Ok(val) = f64::from_str(&input) {
            if val > 0.0 {
                return val;
            }
        }
        println!("Введите положительное число.");
    }
}

fn main() {
    println!("📦 КАЛЬКУЛЯТОР ОБЪЁМА ТЕЛ");
    let shapes = get_shapes();
    let mut history = load_history();
    let mut units = "куб. ед.".to_string();

    loop {
        println!("\nВыберите фигуру:");
        for (key, (name, _, _)) in &shapes {
            println!("{}. {}", key, name);
        }
        println!("h. Показать историю");
        println!("e. Экспорт CSV");
        println!("u. Сменить единицы");
        println!("0. Выход");

        let choice = read_line("Ваш выбор: ");

        if choice == "0" {
            break;
        } else if choice == "h" {
            if history.is_empty() {
                println!("История пуста.");
            } else {
                println!("\n=== ИСТОРИЯ ===");
                let start = if history.len() > 10 { history.len() - 10 } else { 0 };
                for entry in &history[start..] {
                    println!("{} | {} | {}", &entry.date[..19], entry.shape, entry.result);
                }
            }
            continue;
        } else if choice == "e" {
            let filename = read_line("Имя CSV файла (по умолчанию volume_history.csv): ");
            let filename = if filename.is_empty() { "volume_history.csv" } else { &filename };
            export_csv(filename);
            continue;
        } else if choice == "u" {
            println!("Доступные единицы: куб. ед., см³, м³, дм³, л, in³");
            let input = read_line("Выберите единицы: ");
            if !input.is_empty() {
                units = input;
            }
            continue;
        } else if let Some((name, func, param_names)) = shapes.get(&choice) {
            println!("\nФигура: {}", name);
            let mut params = Vec::new();
            for pname in param_names {
                params.push(get_float(&format!("Введите {}: ", pname)));
            }
            let result = func(params.clone());
            let result_str = format!("{:.4} {}", result, units);
            println!("\nОбъём {}: {}", name.to_lowercase(), result_str);

            let save = read_line("Сохранить результат? (y/n): ");
            if save.to_lowercase() == "y" {
                let entry = HistoryEntry {
                    date: chrono::Local::now().to_rfc3339(),
                    shape: name.clone(),
                    params: params.iter().map(|p| p.to_string()).collect::<Vec<_>>().join(", "),
                    result: result_str.clone(),
                    units: units.clone(),
                };
                save_history(&entry);
                history.push(entry);
                println!("✅ Сохранено!");
            }
        } else {
            println!("Неверный выбор.");
        }
    }
}
